tests/data_context/cloud_data_context/test_datasource_crud.py

```diff
@@ -0,0 +1,135 @@
+
+    with patch(
+        "requests.post", autospec=True, side_effect=mocked_post_response
+    ) as mock_post:
+
+        # Call add_datasource with and without the name field included in the datasource config
+        if config_includes_name_setting == "name_supplied_separately":
+            stored_datasource: BaseDatasource = empty_cloud_data_context.add_datasource(
+                name=datasource_name,
+                **datasource_config.to_dict(),
+                save_changes=save_changes,
+            )
+        elif config_includes_name_setting == "config_includes_name":
+            stored_datasource: BaseDatasource = empty_cloud_data_context.add_datasource(
+                **datasource_config_with_name.to_dict(), save_changes=save_changes
+            )
+        elif (
+            config_includes_name_setting
+            == "name_supplied_separately_and_included_in_config"
+        ):
+            stored_datasource: BaseDatasource = empty_cloud_data_context.add_datasource(
+                name=datasource_name,
+                **datasource_config_with_name.to_dict(),
+                save_changes=save_changes,
+            )
+
+        # Make sure we have stored our datasource in the context
+        assert len(empty_cloud_data_context.list_datasources()) == 1
+
+        retrieved_datasource: BaseDatasource = empty_cloud_data_context.get_datasource(
+            datasource_name
+        )
+
+        # This post should have been called without the id (which is retrieved from the response).
+        # It should have been called with the datasource name in the config.
+        if save_changes:
+            mock_post.assert_called_with(
+                f"{ge_cloud_base_url}/organizations/{ge_cloud_organization_id}/datasources",
+                json={
+                    "data": {
+                        "type": "datasource",
+                        "attributes": {
+                            "datasource_config": datasource_config_with_name.to_json_dict(),
+                            "organization_id": ge_cloud_organization_id,
+                        },
+                    }
+                },
+                headers=request_headers,
+            )
+        else:
+            assert not mock_post.called
+
+        if save_changes:
+            # Make sure the id was populated correctly into the created datasource object and config
+            assert stored_datasource.id_ == datasource_id
+            assert retrieved_datasource.id_ == datasource_id
+            assert retrieved_datasource.config["id_"] == datasource_id
+        else:
+            assert stored_datasource.id_ is None
+            assert retrieved_datasource.id_ is None
+            assert retrieved_datasource.config["id_"] is None
+
+        # Make sure the name is populated correctly into the created datasource
+        assert retrieved_datasource.name == datasource_name
+        assert retrieved_datasource.config["name"] == datasource_name
```

Comment:
Nit - might it make sense to split up `save_changes` tests into a separate parameterized test? The logic here can be a bit complex but I feel that's largely personal preference.

Reply:
For sure - I think I got carried away here, just didn't want to duplicate the setup code. I am working on tests for the other data context types in another PR where I'm cleaning this test up & I can split.